### PR TITLE
Regularization API for trainable quantizers

### PR DESCRIPTION
**Pull request by ofirgo**
Extracting the regularization from the quantizer to a higher level.
Now, each trainable quantizer should define its regularization function (if it needs one) near the quantizer implementation, such that the function gets the entire wrapped model and computes the regularization value.
Necessary modifications were applied also to GPTQ configuration and tests.

### CODE REVIEW

File: model_compression_toolkit/gptq/common/gptq_config.py [hunk 1]
```diff
@@ -78,7 +77,7 @@ def __init__(self,
             optimizer_bias (Any): Optimizer to override the rest optimizer for bias.
             log_norm (bool): Whether to use log normalization to the GPTQ Jacobian-based weights.
             weights_n_iter (int): Number of random iterations to run Jacobian approximation for GPTQ weights.
-            quantizer_config (GPTQQuantizerConfig): A class that contains the quantizer specific config.
+            entropy_reg (float): A floating point number that defines the gumbel regularization factor.
```
Comment (elad-c) [inline model_compression_toolkit/gptq/common/gptq_config.py:80]: maybe change to regularization_factor. relevant for all quantizers, not only gumbel

File: model_compression_toolkit/gptq/keras/quantizer/soft_rounding/soft_quantizer_reg.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        return self.start_b * is_before_start_decay + \
+               (1 - is_before_start_decay) * \
+               (self.end_b + (self.start_b - self.end_b) * tf.math.maximum(0.0, (1 - rel_t)))
+
+
+def soft_quantizer_regularization(model: Model, entropy_reg: float, n_batches: int, n_epochs: int) -> float:
+    """
+    Returns the soft quantizer regularization value for SoftRounding.
+
+    Args:
```
Comment (elad-c) [inline model_compression_toolkit/gptq/keras/quantizer/soft_rounding/soft_quantizer_reg.py:72]: missing args

File: model_compression_toolkit/gptq/keras/quantizer/soft_rounding/soft_quantizer_reg.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        is_before_start_decay = tf.cast(t < self.start_decay, tf.float32)
+
+        rel_t = (t - self.start_decay) / (self.t_max - self.start_decay)
+
+        return self.start_b * is_before_start_decay + \
+               (1 - is_before_start_decay) * \
+               (self.end_b + (self.start_b - self.end_b) * tf.math.maximum(0.0, (1 - rel_t)))
+
+
+def soft_quantizer_regularization(model: Model, entropy_reg: float, n_batches: int, n_epochs: int) -> float:
```
Comment (elad-c) [inline model_compression_toolkit/gptq/keras/quantizer/soft_rounding/soft_quantizer_reg.py:68]: can get only total_gragient_steps (n_batches*n_epochs)

File: model_compression_toolkit/gptq/keras/quantizer/soft_rounding/soft_quantizer_reg.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    Args:
+        model: A model to be quantized with SoftRounding.
+        entropy_reg: Entropy value to scale the quantizer regularization.
+
+    Returns: Regularization value.
+    """
+
+    # Initializing the temperature decay according to the number of expected gradient steps
+    init_decay = MAX_ITERATIONS_DEFAULT if n_batches is None else n_epochs * n_batches
```
Comment (elad-c) [inline model_compression_toolkit/gptq/keras/quantizer/soft_rounding/soft_quantizer_reg.py:80]: can change to var name (total_grad...)

File: model_compression_toolkit/gptq/pytorch/quantizer/regularization_factory.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        rounding type, then it returns a function that just returns 0.
+
+    """
+    if gptq_config.rounding_type == RoundingType.SoftQuantizer:
+        # dry run on the representative dataset to count number of batches
+        num_batches = 0
+        for _ in representative_data_gen():
+            num_batches += 1
+
+        n_epochs = N_EPOCHS if not type(gptq_config) == GradientPTQConfigV2 else gptq_config.n_epochs
```
Comment (elad-c) [inline model_compression_toolkit/gptq/pytorch/quantizer/regularization_factory.py:43]: generate GPTQconfigV2 with from_v1. with num_batches as the "n_ptq_iter" argument.
remove N_EPOCHS from code
n_epochs = gptqv2.n_epochs

File: model_compression_toolkit/gptq/pytorch/quantizer/soft_rounding/soft_quantizer_reg.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+        rel_t = (t - self.start_decay) / (self.t_max - self.start_decay)
+
+        return self.start_b * is_before_start_decay + \
+               (1 - is_before_start_decay) * \
+               (self.end_b + (self.start_b - self.end_b) * torch.maximum(to_torch_tensor(np.array([0.0])), (1 - rel_t)))
+
+
+
+def soft_quantizer_regularization(model: nn.Module, entropy_reg: float, n_batches: int, n_epochs: int) -> float:
```
Comment (elad-c) [inline model_compression_toolkit/gptq/pytorch/quantizer/soft_rounding/soft_quantizer_reg.py:70]: dame as keras comments

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 7]
```diff
@@ -66,10 +66,6 @@ def __init__(self,
                                                                        append2output=self.compare_points,
                                                                        fw_info=self.fw_info)
 
-        if self.gptq_config.rounding_type == RoundingType.SoftQuantizer:
-            # dry run on the representative dataset to count number of batches
-            self.count_num_batches_for_training(representative_data_gen)
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/common/gptq_training.py:71]: Remove representative_dataset argument from function (it is not used)

**Review by elad-c — APPROVED** [review]

**Pull request merged**